File: ruby/monosecret_rb/lib/monosecret.rb
# frozen_string_literal: true

# Ruby SDK for Monosecret, a declarative secrets manager.
#
# A thin client over the monosecret_ffi C ABI. The Rust resolver is statically
# linked into a native extension (monosecret_ext), so the SDK inherits every
# provider with no Ruby-side logic and there is nothing to locate at runtime.
# Mirrors the Rust derive crate's vocabulary.

require "json"

# The compiled extension lives next to this file in a source/dev checkout, but in
# an installed gem RubyGems places it in a separate extensions dir already on
# $LOAD_PATH. Put this file's dir on the path so the absolute require resolves in
# both layouts.
$LOAD_PATH.unshift(__dir__) unless $LOAD_PATH.include?(__dir__)
require "monosecret/monosecret_ext"

module Monosecret
  # Response wire-format version this SDK understands. Tracks monosecret_ffi's
  # RESOLVE_SCHEMA_VERSION; a mismatch means the loaded library is incompatible.
  RESOLVE_SCHEMA_VERSION = 2

  # Wire-format version of the value-free report. Tracks monosecret's
  # RESOLUTION_REPORT_SCHEMA_VERSION.
  REPORT_SCHEMA_VERSION = 1

  # A resolution failure (bad manifest, provider error, reason policy).
  class Error < StandardError
    attr_reader :kind

    def initialize(kind, message)
      @kind = kind
      super("#{message} (kind: #{kind})")
    end
  end

  # One or more required secrets were not found anywhere.
  class MissingRequiredError < Error
    attr_reader :missing

    def initialize(missing)
      @missing = missing
      super("missing_required", "missing required secret(s): #{missing.join(', ')}")
    end
  end

  # Caller-asserted software-integration context (SecretSpec 0.20+).
  CallerContext = Struct.new(:name, :version, :operation, :resource, keyword_init: true) do
    def to_h
      { "name" => name, "version" => version, "operation" => operation,
        "resource" => resource }.compact
    end
  end

  # One resolved secret. Exactly one of +value+ / +path+ is set.
  ResolvedSecret = Struct.new(:value, :path, :as_path, :source, :source_provider) do
    # The usable string: the file path for as_path secrets, else the value.
    def get
      as_path ? path : value
    end
  end

  # A successful resolution, mirroring the Rust Resolved wrapper.
  Resolved = Struct.new(:provider, :profile, :secrets, :missing_optional, :scope) do
    # Export each resolved secret into ENV by its declared name. Secrets with no
    # usable value (e.g. under no_values) are skipped rather than deleted from
    # ENV (assigning nil would remove the variable).
    def set_as_env!
      secrets.each do |name, secret|
        value = secret.get
        ENV[name] = value unless value.nil?
      end
    end

    # Flat { "SECRET_NAME" => value } hash (the file path for as_path). A secret
    # with no usable value (e.g. under no_values) maps to nil, matching the null
    # the other SDKs emit. Feed this to a quicktype-generated deserializer (e.g.
    # from_dynamic!). See `monosecret schema`.
    def fields
      secrets.transform_values(&:get)
    end

    # Remove the temp files backing any as_path secrets in this result. The
    # resolver persists those files (mode 0400) so their paths stay valid after
    # resolve returns; the caller owns their lifetime. Call #close (or pass a
    # block to Builder#load, which closes automatically) when done so secret
    # files do not accumulate in the temp dir. A file already gone is not an
    # error.
    #
    # Every file is attempted even if one cannot be removed; the first such
    # error is re-raised once the rest have been cleaned up. Stopping at the
    # first failure would leave the remaining secrets on disk, which is the one
    # outcome this method exists to prevent. Matches the Go SDK's firstErr and
    # the .NET SDK's firstError.
    def close
      first_error = nil
      secrets.each_value do |secret|
        next unless secret.as_path && secret.path

        begin
          File.delete(secret.path)
        rescue Errno::ENOENT
          # already gone
        rescue SystemCallError => e
          first_error ||= e
        end
      end
      raise first_error if first_error

      nil
    end
  end

  # Value-free resolution outcome for one declared secret: how it would resolve
  # and from where, never the value itself.
  SecretReport = Struct.new(:name, :status, :required, :source_provider,
                            :default_applied, :generated, :as_path)

  # A failed cross-secret presence constraint in a resolution report.
  ConstraintViolation = Struct.new(:kind, :group, :secrets, :present)

  # A value-free resolution snapshot. Unlike Resolved, a missing required secret
  # is a "missing_required" status here, not an error, so a report describes a
  # profile even when its secrets are not all available.
  Report = Struct.new(:provider, :profile, :secrets, :scope, :constraint_violations)

  # The narrow C ABI, statically linked into the monosecret_ext extension. The
  # Native.c_resolve / c_abi_version C functions are defined in
  # ext/monosecret/monosecret_ext.c; these wrappers add the Ruby-side error type.
  module Native
    class << self
      def resolve(request_json)
        result = c_resolve(request_json)
        raise Error.new("ffi", "monosecret_resolve returned null") if result.nil?

        result
      end

      def call(request_json)
        unless respond_to?(:c_call, true)
          raise Error.new("capability", "the loaded native extension predates inline specifications; rebuild the secretspec gem")
        end

        result = c_call(request_json)
        raise Error.new("ffi", "secretspec_call returned null") if result.nil?

        result
      end

      def abi_version
        c_abi_version
      end
    end
  end

  # Fluent builder for a resolution.
  class Builder
    def initialize
      @request = {}
      @inline = nil
    end

    def with_path(path)
      @inline = nil
      @request["path"] = path if path
      self
    end

    # Resolve strict inline-spec v1 at its logical base directory (0.20+).
    def with_inline_spec(spec, base_dir)
      @request.delete("path")
      @inline = { "spec" => spec, "base_dir" => base_dir }
      self
    end

    def with_provider(provider)
      @request["provider"] = provider if provider
      self
    end

    def with_profile(profile)
      @request["profile"] = profile if profile
      self
    end

    # Limit resolution to a named manifest scope (Monosecret 0.17+).
    def with_scope(scope)
      @request["scope"] = scope if scope
      self
    end

    def with_reason(reason)
      @request["reason"] = reason if reason
      self
    end

    # Identify the invoking software integration (SecretSpec 0.20+).
    def with_caller(caller)
      @request["caller"] = caller.to_h if caller
      self
    end

    # Omit secret values, returning only structure and provenance.
    def with_no_values(no_values = true)
      @request["no_values"] = no_values
      self
    end

    # Resolve the secrets. Raises MissingRequiredError if a required secret is
    # missing, and Error for any other failure.
    #
    # Without a block, returns the Resolved (the caller should #close it when
    # done to clean up any as_path temp files). With a block, yields the Resolved
    # and closes it afterwards, returning the block's value.
    def load
      response = parse_response(*native_request, "resolve", RESOLVE_SCHEMA_VERSION)

      missing = response["missing_required"] || []
      raise MissingRequiredError.new(missing) unless missing.empty?

      secrets = {}
      (response["secrets"] || {}).each do |name, entry|
        secrets[name] = ResolvedSecret.new(
          entry["value"], entry["path"], entry["as_path"] || false,
          entry["source"], entry["source_provider"]
        )
      end

      resolved = Resolved.new(
        response["provider"], response["profile"], secrets,
        response["missing_optional"] || [], response["scope"]
      )
      return resolved unless block_given?

      begin
        yield resolved
      ensure
        resolved.close
      end
    end

    # Resolve a value-free Report (the inventory/preflight view, the same one the
    # CLI exposes as `check --json`). Unlike #load, never raises
    # MissingRequiredError: a missing required secret appears as a SecretReport
    # with status "missing_required".
    def report
      response = parse_response(*native_request("report"), "report", REPORT_SCHEMA_VERSION)

      secrets = (response["secrets"] || []).map do |s|
        SecretReport.new(s["name"], s["status"], s["required"],
                         s["source_provider"], s["default_applied"],
                         s["generated"], s["as_path"])
      end
      violations = (response["constraint_violations"] || []).map do |violation|
        ConstraintViolation.new(violation["kind"], violation["group"],
                                violation["secrets"], violation["present"])
      end
      Report.new(response["provider"], response["profile"], secrets,
                 response["scope"], violations)
    end

    private

    # Resolve a JSON request payload and return the validated "response" hash, or
    # raise. +kind+ is "resolve" or "report"; it selects the schema version to
    # enforce and labels the version-mismatch message.
    def parse_response(request, versioned, kind, expected_version)
      payload = JSON.generate(request)
      envelope = JSON.parse(versioned ? Native.call(payload) : Native.resolve(payload))

      unless envelope["ok"]
        err = envelope["error"] || {}
        raise Error.new(err["kind"] || "unknown", err["message"] || "")
      end

      response = envelope["response"]
      raise Error.new("ffi", "monosecret_resolve reported ok with no response") if response.nil?

      version = response["schema_version"]
      unless version == expected_version
        raise Error.new("version",
                        "unsupported #{kind} schema version #{version} " \
                        "(expected #{expected_version}); the monosecret_ffi " \
                        "library and this SDK are out of sync")
      end

      response
    end

    def native_request(mode = nil)
      options = @request.dup
      options["mode"] = mode if mode
      return [options, false] unless @inline

      [{ "request_version" => 1, "operation" => "resolve",
         "source" => { "kind" => "inline", "spec_version" => 1,
                       "base_dir" => @inline["base_dir"], "spec" => @inline["spec"] },
         "options" => options }, true]
    end
  end

  def self.builder
    Builder.new
  end

  def self.abi_version
    Native.abi_version
  end
end

File: haskell/monosecret_hs/src/Monosecret.hs
{-# LANGUAGE ForeignFunctionInterface #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Haskell SDK for Monosecret, a declarative secrets manager.
--
-- A thin client over the @monosecret_ffi@ C ABI, linked at build time.
-- Resolution (providers, fallback chains, profiles, generation, @as_path@)
-- happens entirely in the Rust core; this module marshals a JSON request to
-- @monosecret_resolve@, parses the response envelope, and exposes it with the
-- same vocabulary as the Rust derive crate.
--
-- > import qualified Monosecret as S
-- > import Data.Function ((&))
-- >
-- > main = do
-- >   resolved <- S.load (S.builder & S.withProvider "keyring://" & S.withReason "boot")
-- >   print (S.get =<< Data.Map.lookup "DATABASE_URL" (S.resolvedSecrets resolved))
-- >   S.setAsEnv resolved
module Monosecret
  ( -- * Builder
    Builder
  , CallerContext(..)
  , builder
  , withPath
  , withInlineSpec
  , withProvider
  , withProfile
  , withScope
  , withReason
  , withCaller
  , withNoValues
    -- * Resolve (value-carrying)
  , Resolved(..)
  , ResolvedSecret(..)
  , load
  , get
  , fields
  , fieldsJson
  , setAsEnv
  , close
    -- * Report (value-free)
  , Report(..)
  , SecretReport(..)
  , ConstraintViolation(..)
  , ConstraintViolationKind(..)
  , report
    -- * Errors
  , MonosecretError(..)
  , MissingRequiredError(..)
    -- * Misc
  , abiVersion
  ) where

import           Control.Exception (Exception, finally, mask, throwIO)
import           Control.Monad (forM_, unless, when)
import           Data.Aeson (FromJSON (..), Value, eitherDecodeStrict, encode,
                             object, withObject, withText, (.!=), (.:), (.:?), (.=))
import           Data.Aeson.Types (parseEither)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import           Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import           Data.Maybe (catMaybes)
import           Data.Text (Text)
import qualified Data.Text as T
import           Foreign.C.String (CString, peekCString)
import           Foreign.Ptr (nullPtr)
import           System.Directory (doesFileExist, removeFile)
import qualified System.Environment.Blank as Env

-- The three C ABI functions, linked at build time. The default build embeds the
-- static archive; pkg-config builds may use a static or shared install. They
-- are declared @safe@ because @monosecret_resolve@ may block on provider I/O
-- (1Password, LastPass), and a @safe@ call lets other Haskell threads run.
foreign import ccall safe "monosecret_resolve"
  c_monosecret_resolve :: CString -> IO CString

foreign import ccall safe "monosecret_call"
  c_monosecret_call :: CString -> IO CString

foreign import ccall safe "monosecret_free"
  c_monosecret_free :: CString -> IO ()

foreign import ccall safe "monosecret_abi_version"
  c_monosecret_abi_version :: IO CString

-- | Wire-format version of the value-carrying resolve response this SDK
-- understands. Tracks @monosecret@'s @RESOLVE_SCHEMA_VERSION@.
resolveSchemaVersion :: Int
resolveSchemaVersion = 2

-- | Wire-format version of the value-free report. Tracks @monosecret@'s
-- @RESOLUTION_REPORT_SCHEMA_VERSION@.
reportSchemaVersion :: Int
reportSchemaVersion = 1

-- | A resolution failure (bad manifest, provider error, reason policy). Carries
-- a stable @kind@.
data MonosecretError = MonosecretError
  { errorKind    :: Text
  , errorMessage :: Text
  } deriving (Show, Eq)

instance Exception MonosecretError

-- | One or more required secrets were not found anywhere.
newtype MissingRequiredError = MissingRequiredError
  { missing :: [Text]
  } deriving (Show, Eq)

instance Exception MissingRequiredError

-- | One resolved secret. Exactly one of 'secretValue' \/ 'secretPath' is set:
-- the path for @as_path@ secrets, the value otherwise. Both are 'Nothing' for a
-- value-less ('withNoValues') response.
data ResolvedSecret = ResolvedSecret
  { secretValue          :: Maybe Text
  , secretPath           :: Maybe Text
  , secretAsPath         :: Bool
  , secretSource         :: Text
  , secretSourceProvider :: Maybe Text
  } deriving (Show, Eq)

instance FromJSON ResolvedSecret where
  parseJSON = withObject "ResolvedSecret" $ \o ->
    ResolvedSecret
      <$> o .:? "value"
      <*> o .:? "path"
      <*> o .:? "as_path" .!= False
      <*> o .: "source"
      <*> o .:? "source_provider"

-- | A successful resolution, mirroring the Rust @Resolved@ wrapper.
data Resolved = Resolved
  { resolvedProvider        :: Text
  , resolvedProfile         :: Text
  -- | Selected manifest scope, or 'Nothing' for a full-profile resolve (0.17+).
  , resolvedScope           :: Maybe Text
  , resolvedSecrets         :: Map Text ResolvedSecret
  , resolvedMissingOptional :: [Text]
  } deriving (Show, Eq)

-- | The value-free resolution outcome for one declared secret: how it would
-- resolve and from where, never the value itself.
data SecretReport = SecretReport
  { srName           :: Text
  , srStatus         :: Text -- ^ @"resolved"@, @"missing_required"@, or @"missing_optional"@.
  , srRequired       :: Bool
  , srSourceProvider :: Maybe Text
  , srDefaultApplied :: Bool
  , srGenerated      :: Bool
  , srAsPath         :: Bool
  } deriving (Show, Eq)

instance FromJSON SecretReport where
  parseJSON = withObject "SecretReport" $ \o ->
    SecretReport
      <$> o .: "name"
      <*> o .: "status"
      <*> o .:? "required" .!= False
      <*> o .:? "source_provider"
      <*> o .:? "default_applied" .!= False
      <*> o .:? "generated" .!= False
      <*> o .:? "as_path" .!= False

-- | The kind of a failed cross-secret presence constraint.
data ConstraintViolationKind = AtLeastOne | ExactlyOne
  deriving (Show, Eq)

instance FromJSON ConstraintViolationKind where
  parseJSON = withText "ConstraintViolationKind" $ \kind ->
    case kind of
      "at_least_one" -> pure AtLeastOne
      "exactly_one" -> pure ExactlyOne
      _ -> fail ("unknown constraint violation kind: " ++ T.unpack kind)

-- | A failed cross-secret presence constraint in a resolution report.
data ConstraintViolation = ConstraintViolation
  { violationKind    :: ConstraintViolationKind
  , violationGroup   :: Text
  , violationSecrets :: [Text]
  , violationPresent :: [Text]
  } deriving (Show, Eq)

instance FromJSON ConstraintViolation where
  parseJSON = withObject "ConstraintViolation" $ \o ->
    ConstraintViolation
      <$> o .: "kind"
      <*> o .: "group"
      <*> o .: "secrets"
      <*> o .: "present"

-- | A value-free resolution snapshot. Unlike 'Resolved', a missing required
-- secret is a @"missing_required"@ status here, not an error.
data Report = Report
  { reportProvider             :: Text
  , reportProfile              :: Text
  -- | Selected manifest scope, or 'Nothing' for a full-profile report (0.17+).
  , reportScope                :: Maybe Text
  , reportSecrets              :: [SecretReport]
  , reportConstraintViolations :: [ConstraintViolation]
  } deriving (Show, Eq)

-- | A resolution request. Build it from 'builder' with the @withX@ setters and
-- pass it to 'load' or 'report'.
data Builder = Builder
  { bPath     :: Maybe Text
  , bProvider :: Maybe Text
  , bProfile  :: Maybe Text
  , bScope    :: Maybe Text
  , bReason   :: Maybe Text
  , bCaller   :: Maybe CallerContext
  , bNoValues :: Bool
  , bInline   :: Maybe (Value, Text)
  }

-- | Caller-asserted software-integration context (Monosecret 0.20+). It is
-- audit metadata and never supplies the user access reason.
data CallerContext = CallerContext
  { callerName      :: Text
  , callerVersion   :: Maybe Text
  , callerOperation :: Maybe Text
  , callerResource  :: Maybe Text
  } deriving (Show, Eq)

-- | A builder with no options set.
builder :: Builder
builder = Builder Nothing Nothing Nothing Nothing Nothing Nothing False Nothing

-- | Resolve from a manifest at this path instead of walking up from the working
-- directory.
withPath :: Text -> Builder -> Builder
withPath v b = b { bPath = Just v, bInline = Nothing }

-- | Resolve strict inline-spec v1 at its logical base directory (0.20+).
-- The static linker requires @monosecret_call@, so an older native archive
-- fails at link time instead of falling back to a filesystem manifest.
withInlineSpec :: Value -> Text -> Builder -> Builder
withInlineSpec spec baseDir b = b { bPath = Nothing, bInline = Just (spec, baseDir) }

-- | Override the provider (a @keyring:\/\/@-style URI or a configured alias).
withProvider :: Text -> Builder -> Builder
withProvider v b = b { bProvider = Just v }

-- | Override the profile.
withProfile :: Text -> Builder -> Builder
withProfile v b = b { bProfile = Just v }

-- | Limit resolution to a named manifest scope (Monosecret 0.17+).
withScope :: Text -> Builder -> Builder
withScope v b = b { bScope = Just v }

-- | Set a human-readable reason for this access (for audited providers).
withReason :: Text -> Builder -> Builder
withReason v b = b { bReason = Just v }

-- | Identify the invoking software integration (Monosecret 0.20+).
withCaller :: CallerContext -> Builder -> Builder
withCaller v b = b { bCaller = Just v }

-- | Omit secret values, returning only structure and provenance.
withNoValues :: Bool -> Builder -> Builder
withNoValues v b = b { bNoValues = v }

-- | The usable string: the file path for @as_path@ secrets, else the value.
-- 'Nothing' when no usable value is present (e.g. under 'withNoValues').
get :: ResolvedSecret -> Maybe Text
get s = if secretAsPath s then secretPath s else secretValue s

-- | Flat @name -> usable value@ map ('Nothing' encodes to JSON @null@), the
-- input for a quicktype-generated deserializer. See @monosecret schema@.
fields :: Resolved -> Map Text (Maybe Text)
fields = Map.map get . resolvedSecrets

-- | 'fields' as a JSON byte string (a @{SECRET_NAME: value-or-null}@ object).
fieldsJson :: Resolved -> BL.ByteString
fieldsJson = encode . fields

-- | Export each resolved secret into the process environment by its declared
-- name. Secrets with no usable value (e.g. under 'withNoValues') are skipped.
setAsEnv :: Resolved -> IO ()
setAsEnv r =
  forM_ (Map.toList (resolvedSecrets r)) $ \(name, secret) ->
    case get secret of
      -- System.Environment.setEnv treats setEnv name "" as unsetEnv name, which
      -- would *delete* a secret that resolves to "" (e.g. `.env` line `FOO=` or
      -- `default = ""`). System.Environment.Blank.setEnv with overwrite=True sets
      -- the empty string, matching the Go/Python/Ruby/Node SDKs.
      Just v  -> Env.setEnv (T.unpack name) (T.unpack v) True
      Nothing -> pure ()

-- | Remove the temp files backing any @as_path@ secrets in this result. The
-- resolver persists those files (mode 0400) so their paths stay valid after
-- resolve returns; the caller owns their lifetime. Call 'close' when done so the
-- secret files do not accumulate in the temp dir. A file already gone is ignored.
close :: Resolved -> IO ()
close r =
  forM_ (Map.elems (resolvedSecrets r)) $ \secret ->
    case (secretAsPath secret, secretPath secret) of
      (True, Just p) -> do
        let fp = T.unpack p
        exists <- doesFileExist fp
        when exists (removeFile fp)
      _ -> pure ()

-- | The ABI version reported by the loaded native library.
abiVersion :: IO Text
abiVersion = do
  -- A static, library-owned string; do not free it.
  c <- c_monosecret_abi_version
  T.pack <$> peekCString c

-- | Resolve the secrets. Throws 'MissingRequiredError' if a required secret is
-- missing, and 'MonosecretError' for any other failure.
load :: Builder -> IO Resolved
load b = do
  resp <- callNative (isInline b) (requestBytes b Nothing)
  value <- responseValue resp resolveSchemaVersion "resolve"
  (prov, prof, scope, secs, mreq, mopt) <- fromResult (parseEither pResolve value)
  case mreq of
    [] -> pure (Resolved prov prof scope secs mopt)
    xs -> throwIO (MissingRequiredError xs)
  where
    pResolve = withObject "response" $ \o ->
      (,,,,,)
        <$> o .: "provider"
        <*> o .: "profile"
        <*> o .:? "scope"
        <*> o .:? "secrets" .!= Map.empty
        <*> o .:? "missing_required" .!= []
        <*> o .:? "missing_optional" .!= []

-- | Resolve a value-free 'Report' (the inventory\/preflight view, the same one
-- the CLI exposes as @check --json@). Unlike 'load', it does not throw when a
-- required secret is missing: that secret appears as a 'SecretReport' with
-- status @"missing_required"@.
report :: Builder -> IO Report
report b = do
  resp <- callNative (isInline b) (requestBytes b (Just "report"))
  value <- responseValue resp reportSchemaVersion "report"
  (prov, prof, scope, secs, violations) <- fromResult (parseEither pReport value)
  pure (Report prov prof scope secs violations)
  where
    pReport = withObject "response" $ \o ->
      (,,,,)
        <$> o .: "provider"
        <*> o .: "profile"
        <*> o .:? "scope"
        <*> o .:? "secrets" .!= []
        <*> o .:? "constraint_violations" .!= []

-- Build the request JSON for a resolve (@mode = Nothing@) or report
-- (@mode = Just "report"@), omitting unset options.
requestBytes :: Builder -> Maybe Text -> BL.ByteString
requestBytes b mode =
  case bInline b of
    Nothing -> encode options
    Just (spec, baseDir) -> encode $ object
      [ "request_version" .= (1 :: Int)
      , "operation" .= ("resolve" :: Text)
      , "source" .= object
          [ "kind" .= ("inline" :: Text)
          , "spec_version" .= (2 :: Int)
          , "base_dir" .= baseDir
          , "spec" .= spec
          ]
      , "options" .= options
      ]
  where
    options = object $
      catMaybes
        [ ("path" .=) <$> bPath b
        , ("provider" .=) <$> bProvider b
        , ("profile" .=) <$> bProfile b
        , ("scope" .=) <$> bScope b
        , ("reason" .=) <$> bReason b
        , ("caller" .=) . callerValue <$> bCaller b
        ]
        ++ ["no_values" .= True | bNoValues b]
        ++ ["mode" .= m | Just m <- [mode]]
    callerValue caller = object . catMaybes $
      [ Just ("name" .= callerName caller)
      , ("version" .=) <$> callerVersion caller
      , ("operation" .=) <$> callerOperation caller
      , ("resource" .=) <$> callerResource caller
      ]

-- Marshal a request to monosecret_resolve and copy the response out before
-- freeing the native allocation.
--
-- The response is a Rust allocation the caller must free, and it carries secret
-- values. @mask@ keeps an async exception (e.g. a 'System.Timeout.timeout'
-- around 'load') from landing between the call returning and the free being
-- installed, and @finally@ guarantees the free runs whether @packCString@
-- succeeds, throws, or is interrupted — so the secret-bearing buffer never leaks.
isInline :: Builder -> Bool
isInline = maybe False (const True) . bInline

callNative :: Bool -> BL.ByteString -> IO BS.ByteString
callNative versioned reqLazy =
  BS.useAsCString (BL.toStrict reqLazy) $ \creq ->
    mask $ \restore -> do
      cresp <- (if versioned then c_monosecret_call else c_monosecret_resolve) creq
      if cresp == nullPtr
        then throwIO (MonosecretError "ffi" (if versioned then "monosecret_call returned null" else "monosecret_resolve returned null"))
        else restore (BS.packCString cresp) `finally` c_monosecret_free cresp

-- Decode the envelope, unwrap @ok@/@error@, and check the schema version,
-- returning the response object as a 'Value' for the caller to project.
responseValue :: BS.ByteString -> Int -> Text -> IO Value
responseValue resp expectVer kind = do
  env <- case eitherDecodeStrict resp :: Either String (Envelope Value) of
    Left e  -> throwIO (MonosecretError "parse" (T.pack e))
    Right v -> pure v
  if not (envOk env)
    then case envError env of
      Just (ErrInfo k m) -> throwIO (MonosecretError k m)
      Nothing            -> throwIO (MonosecretError "unknown" "")
    else case envResponse env of
      Nothing -> throwIO (MonosecretError "ffi" "monosecret_resolve reported ok with no response")
      Just value -> do
        ver <- fromResult (parseEither (withObject "response" (.: "schema_version")) value)
        unless (ver == expectVer) (throwIO (versionError ver expectVer kind))
        pure value

versionError :: Int -> Int -> Text -> MonosecretError
versionError got expected kind =
  MonosecretError "version" $
    T.concat
      [ "unsupported ", kind, " schema version ", T.pack (show got)
      , " (expected ", T.pack (show expected)
      , "); the monosecret_ffi library and this SDK are out of sync"
      ]

fromResult :: Either String a -> IO a
fromResult = either (throwIO . MonosecretError "parse" . T.pack) pure

-- The response envelope shared by every native binding.
data Envelope a = Envelope
  { envOk       :: Bool
  , envResponse :: Maybe a
  , envError    :: Maybe ErrInfo
  }

instance FromJSON a => FromJSON (Envelope a) where
  parseJSON = withObject "Envelope" $ \o ->
    Envelope <$> o .: "ok" <*> o .:? "response" <*> o .:? "error"

data ErrInfo = ErrInfo Text Text

instance FromJSON ErrInfo where
  parseJSON = withObject "error" $ \o -> ErrInfo <$> o .: "kind" <*> o .: "message"
